pkg/generate/app/cmd/newapp.go

```diff
@@ -717,13 +717,86 @@ func (c *AppConfig) checkCircularReferences(objects app.Objects) error {
 	for _, obj := range objects {
 		if bc, ok := obj.(*buildapi.BuildConfig); ok {
 			input := buildutil.GetInputReference(bc.Spec.Strategy)
-			if bc.Spec.Output.To != nil && input != nil &&
-				reflect.DeepEqual(input, bc.Spec.Output.To) {
-				ns := input.Namespace
-				if len(ns) == 0 {
-					ns = c.OriginNamespace
+			output := bc.Spec.Output.To
+
+			if output == nil || input == nil {
+				return nil
+			}
+
+			inNS := input.Namespace
+			if len(inNS) == 0 {
+				inNS = c.OriginNamespace
+			}
+
+			// We might be outputing a docker image that would trigger our imagestream repo.
+			if output.Kind == "DockerImage" && input.Kind == "ImageStreamTag" {
+				isName, isTag, ok := imageapi.SplitImageStreamTag(input.Name)
+				if !ok {
+					return fmt.Errorf("Unable to parse ImageStreamTag name: %q", input.Name)
 				}
-				return app.CircularOutputReferenceError{Reference: fmt.Sprintf("%s/%s", ns, input.Name)}
+
+				// The imagestream is usually being created alongside the buildconfig
+				// when new-build is being used, so scan objects being created for it.
+				var is *imageapi.ImageStream
+				for _, check := range objects {
+					if is2, ok := check.(*imageapi.ImageStream); ok {
+						if is2.Name == isName {
+							is = is2
+							break
+						}
+					}
+				}
+
+				if is == nil {
+					var err error
+					is, err = c.OSClient.ImageStreams(inNS).Get(isName)
+					if err != nil {
+						glog.Warningf("Unable to check for circular build input/outputs")
+						return nil
+					}
+				}
+
+				// Dereference ImageStream so that we compare DockerImage against DockerImage.
+				input = is.Spec.Tags[isTag].From
+
+				if input == nil {
+					if is.Spec.DockerImageRepository == "" {
+						glog.Warningf("Unable to check for circular build input/outputs")
+						return nil
+					}
+					// Legacy InputStream without tag support? Spoof what we should see.
+					input = &kapi.ObjectReference{
+						Kind: "DockerImage",
+						Name: is.Spec.DockerImageRepository,
+					}
+				}
+			}
+
+			// For DeepEqual to work for all input variations, we need to massage the
+			// data a bit. Make copies so we don't affect actual data structures.
+			inCopy := *input
+			outCopy := *output
+
+			for _, ref := range []*kapi.ObjectReference{&inCopy, &outCopy} {
+				if ref.Kind == "DockerImage" {
```

Comment:
@jupierce would it make sense to add some of this commentary to the code itself?
Not the comments replace clear code, but perhaps it will help us better if more people can understand what is going on there, what the corner cases are.

Having good tests also helps.